proposals/windows-mdm-commands-tables.md

```diff
@@ -0,0 +1,118 @@
+# Changes to Windows MDM command tables
+
+**Context** Commands are the mechanism we'll use to build features like
+
+- Deliver OS settings
+- Filter hosts by their "OS settings" status (pending, verified, verifying, failed)
+
+IF possible I would like us to have the same mental model for both platforms,
+where we have three tables with a clear separation of concerns:
+
+A. List of commands
+B. Queue
+C. List of command results
+
+Any command, doesn't matter how it was enqueued (via the UI/CLI or in response
+to other commands) should have an entry in `A`, plus entries with FK references
+in `B` when it's enqueued, and `C` when we get results.
+
+The [current schema](https://github.com/fleetdm/fleet/blob/defad14dd92ac1fb049552ec1564bceb50606e2d/server/datastore/mysql/migrations/tables/20230926082152_AddWindowsMDMTables.go)
+for Windows has two tables, with the following descriptions:
+
+### `windows_mdm_pending_commands`
+
+> The table windows_mdm_pending_commands contains the stateless commands enqueued into the MDM pending commands table from fleetctl or fleet Server.
+> These commands contain operations to be performed on a given device. 
+> They don't yet have the MessageID, SessionID, and CmdId information as this data gets populated during the response 
+> to a given MDM management request.
+
+### `windows_mdm_commands`
+
+> The windows_mdm_commands table represents the stateful commands that were actually sent over the wire. 
+> These commands include the MessageID, SessionID, and CmdId as they were sent to the device. 
+> They are stored in the commands table for error code tracking purposes. 
+> Additionally, this table can contain commands that were not part of the windows_mdm_pending_commands table and 
+> were generated by the Fleet server in response to an incoming MDM command.
+
+## Proposal
+
+I think it's fundamental we have:
+
+- A clear queue model
+- A single source of truth for commands
+- The status of a command at any point in time easily available
+- The raw XML we sent to the device
+- The raw XML we received as a response
+
+A rough version of the schema (many missing fields from the original) could be:
+
+```sql
+CREATE TABLE windows_mdm_commands (
+    -- managed and generated by Fleet
+    command_uuid    VARCHAR(127) NOT NULL,
+
+    -- raw XML
+    raw_command      MEDIUMTEXT   NOT NULL,
+
+    -- command_id specifies a unique command identifier for the SyncML message (Section 2.2.3.2 on MS-MDM spec)
+    command_id      VARCHAR(255) NOT NULL,
```

Comment:
We may not have that field when receiving commands to run from the users. I don't know if we can just generate it, but if not it might need to be nullable (or empty) for those cases, or we should maybe have it only in the results table, not sure?

Reply:
> If we go with the approach of defining the command_id upfront and make it to be a UUID, we can remove command_id and use command_uuid as the CommandID of the protocol operation.

:+1:  I think that's a huge benefit.